### ISSUE TITLE
Running all JUnit tests in a project fails for Flatpak Eclipse

### ISSUE THREAD

**Issue by syphr42** [flathub/org.eclipse.Java#42]
I asked this as a question over on [StackOverflow](https://stackoverflow.com/questions/68154020/how-do-you-run-junit-tests-in-flatpak-eclipse), but I'm starting to think it is a bug with the Flatpak version of Eclipse rather than a configuration problem. Running the same unit tests in non-Flatpak Eclipse works fine. I'm on the latest version (2021-06).

When I try to run a JUnit 4 suite defined to run all tests within a project, it fails with this exception:

```
java.lang.IllegalArgumentException: Cannot read testname file.
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.defaultInit(RemoteTestRunner.java:266)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.init(RemoteTestRunner.java:225)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:209)
```

I looked into the source for `RemoteTestRunner` and found that it is trying to read a file whose path was provided to the runner:

```
			else if("-testnamefile".equals(args[i].toLowerCase())) { //$NON-NLS-1$
				String testNameFile= args[i+1];
				try {
					readTestNames(testNameFile);
				} catch (IOException e) {
					throw new IllegalArgumentException("Cannot read testname file.");		 //$NON-NLS-1$
				}
				i++;

			}
```
[Source link](https://github.com/eclipse/eclipse.jdt.ui/blob/724f841f19160133a493dfb33d36ee550b219857/org.eclipse.jdt.junit.runtime/src/org/eclipse/jdt/internal/junit/runner/RemoteTestRunner.java#L261-L269)

So I examined the command line for the test execution and did find that argument:
```
-testNameFile /tmp/testNames8297875472451881369.txt
```

The exception message isn't lying - that file does not exist. The name of the file changes each time it is run, but the file is never created.

I verified from a terminal inside Eclipse and from the Flatpak container via `flatpak run --command=sh org.eclipse.Java` that I can create, modify, and delete files at `/tmp`.

This is where I'm stuck. I'm not sure what is supposed to create this temporary file with the test names in it or why it is failing to do so. I see no evidence in Eclipse logs of a failure to write this file. Permissions don't seem to be an issue.

One last observation I did make is that when I open a terminal inside Eclipse and run `ls -al /tmp` I see all the files from my host machine's `/tmp` directory. I thought Flatpak specifically blocked `/tmp` [even when file permissions are set to host](https://docs.flatpak.org/en/latest/sandbox-permissions.html#filesystem-access). However, when I run the sandbox from a terminal (using `flatpak run`), I see a sandbox version of `/tmp` that does not match the host. In either case, I can read, write, and modify files in `/tmp`.

**Comment by akurtakov** [comment]
@vrubezhny would you please investigate?

**Comment by vrubezhny** [comment]
> @vrubezhny would you please investigate?

Sure.

**Comment by vrubezhny** [comment]
@syphr42 According to https://docs.flatpak.org/en/latest/sandbox-permissions.html#filesystem-access:

> The main rules are:
> 
> These directories are blacklisted: /lib, /lib32, /lib64, /bin, /sbin, /usr, /boot, /root, /tmp, /etc, /app, /run, /proc, /sys, /dev, /var
> Exceptions from the blacklist: /run/media
> These directories are mounted under /var/run/host: /etc, /usr
> 

So, I think JUnit cannot write/read anything at `/tmp`. But probably i's possible to  add a wrapper script that maps `/tmp` to a directory inside  `$XDG_RUNTIME_DIR/app/$FLATPAK_ID` (f.i. to `$XDG_RUNTIME_DIR/app/$FLATPAK_ID/tmp`).

**Comment by vrubezhny** [comment]
@syphr42 As more simple workaround you could probably:
- create a "temp" directory somewhere inside your home directory or your workspace;
- add `-Djava.io.tmpdir="<put here the full path to your temp directory>"` (for example, `-Djava.io.tmpdir="/home/user/temp/"`) to `VM Arguments` on `Arguments` page of your Run/Debug Launch configuration wizard.

This will make JUnit to use your "temp" directory instead of default `/tmp` directory.

Could you please give some details on your configuration please, like version of your OS?

**Comment by akurtakov** [comment]
@vrubezhny this is flatpak installation (see https://flathub.org/apps/details/org.eclipse.Java). You may have not used it yet but OS is irrelevant as it's sandbox technology that's same on every OS. Please install from flathub and try debugging the code to familiarize yourself. IMHO it's the future for Eclipse IDE distribution on linux.

**Comment by vrubezhny** [comment]
@akurtakov Asking because I can't reproduce the problem and it looks like `/tmp` directory is available for read/write for my 'org.eclipse.Java` installation  - that's why I'm asking.

**Comment by syphr42** [comment]
I am able to confirm that adding the VM arg `-Djava.io.tmpdir` to the arguments for my JUnit test runner configuration with a directory inside my home directory did not work. I still receive the same error. I also agree @vrubezhny that Flatpak docs say that `/tmp` should not be accessible, but my experimentation seems to show the same thing you found and that is that `/tmp` is accessible. This of course does not help explain why Eclipse can't write to it.

In any case, I see that there is a linked pull request now. I will continue the discussion over on the PR.

**Comment by vrubezhny** [comment]
@syphr42 Actually, `/tmp` directory is still active in a sandbox, but it's not the same as normal `/tmp` filesystem directory. It looks like it's bieng redirected to something like `/dev/null` - so it can be written to without any error indication (technically it can be read from, but there is nothing to read). As such `testNames.txt`temporary file is created and is written without any error indication and then immediately `disappeared`.

As far as I understand we're still allowed to override this behavior - the PR is a try to allow access to filesystem's `/tmp` directory from inside the sandbox.

**Comment by syphr42** [comment]
This is resolved by #43

Thanks for the quick fix on this!
